first attempt :)

freedomride opened the pull request.

freedomride committed:
first attempt :)

s1dney commented on dna/Cell.m:
```diff
@@ -0,0 +1,64 @@
+        _setOfDNA = @[@"A", @"T", @"G", @"C"];
+        
+        do {
+            [_DNA addObject:[_setOfDNA objectAtIndex:(arc4random() % 3)]];
```
arc4random () % 3 генерирует рандомное число и возвращает остаток от деления числа на 3, соответственно это 0, 1 или 2
таким образом в массиве никогда не будет символа С, у которого индекс в массиве 3

freedomride committed:
fix random

s1dney commented on dna/main.m:
```diff
@@ -0,0 +1,63 @@
+        NSLog(@"%i",[dna1 hammingDistance:dna2]);
+        
+        [dna1 mutate:33];
+        [dna1 mutate:60];
```
нужно мутировать оба объекта, здесь же мутирует только 1, но 2 раза

s1dney commented on dna/main.m:
```diff
@@ -0,0 +1,63 @@
+    
+    float countIndexesForNotSuffle = lengthOfDNA - round(lengthOfDNA/100.0*percentForSuffle);
+    for (int i=0; i<countIndexesForNotSuffle; i++) {
+        [indexesForSuffle removeObjectAtIndex:(arc4random() % (lengthOfDNA - i))];
```
подход неверный, в этом цикле может 5 раз подряд удалиться значение с индексом 0, например, соответственно в массиве останется больше значений, чем этого требуется для мутации

s1dney commented on dna/main.m:
```diff
@@ -0,0 +1,63 @@
+        [setOfDNAWithoutCurrent addObjectsFromArray:self.setOfDNA];
+        [setOfDNAWithoutCurrent removeObject:[self.DNA objectAtIndex:[[indexesForSuffle objectAtIndex:i] integerValue]]];
+        
+        [self.DNA replaceObjectAtIndex:[[indexesForSuffle objectAtIndex:i] integerValue] withObject:[setOfDNAWithoutCurrent objectAtIndex:(arc4random() % 3)]];
```
если честно, то я просто нихрена не понял =) вот как я это вижу:
создается массив из 4 значений
потом очищается
потом опять добавляются 4 значения
затем удаляется значение из интервала от 1 до 100 (если DNA 100) при том, что массив всего из 4 элементов
и наконец элемент массива DNA заменяется с рандомным символом из малого массива
опять же arc4random() %3 
но главное, что условия задачи не выполнены: старый символ и новый могут совпасть и мутации не произойдет

s1dney commented on dna/main.m:
```diff
@@ -0,0 +1,63 @@
+    
+    NSMutableArray *indexesForSuffle = [NSMutableArray arrayWithCapacity:lengthOfDNA];
+    for (int i=0; i<lengthOfDNA; i++) {
+        [indexesForSuffle addObject:[NSString stringWithFormat:@"%d",i]];
```
заполнение массива происходит ленейно - 0 1 2 3 4, и мутация соответственно тоже
по условиям задачи нужно производить мутацию по случайному индексу

aleksey-ios-dev commented:
Только учусь читать чужой код, пока плаваю... но здесь я что-то вообще не могу врубиться, что происходит...

freedomride committed:
fix in name of varible for mutate

Спасибо s1dney за выявленные ошибки)
И спасибо всем, кто пытается проверить мою работу)
Добавил более подробное описания как всё работает.
Также добавил много NSLog'ов, чтоб легче было проверять и понять)

freedomride commented:
Спасибо @s1dneyза выявленные ошибки)
И спасибо всем, кто пытается проверить мою работу)

to @s1dney 
постарался более детально расписать процесс)
по поводу цикла которые может много раз удалить индекс 0:
изначально массив имеет вид @[@"0",@"1",@"2",@"3"...@"99"]
если из него удалить например два раза позицию по индексу 0 то получим @[@"2",@"3"...@"99"]
по поводу массива из 4-х значений:
изначально он пустой, далее не совсем корректно было написано что пустой еще раз очищаем, поправил (перенёс в конец), но в принципе это не на что не влияем
далее мы его заполняем массивом всех возможных типов нуклеотидов и удаляем из него текущий. итого имеем массив типов нуклеотидов без текущего.
по остальным попытался в коде комментариями пояснить.

to @mmrmmlrr 
добавил по больше комментариев, надеюсь поможет)

s1dney commented:
ну вот смотри:
ты рассчитал процент, например это 30 от 100, создается цикл на удаление лишних 70 значений
первый раз он удалил индекс 0
второй раз тоже 0
и третий раз 0, а счетчик в это время там считает i++, хоть по факту ничего и не удалилось
если даже предположить,что больше повторений не было, то получается, что у тебя в место 70 значений удалилось только 68

самый простой способ это проверить - сделай NSLog после этого цикла и ты увидишь, что там элементов больше чем нужно 

вообще, как оказалось, arc4random очень злая вещь, нужно быть осторожнее с ним, сам на эти же грабли наступал

s1dney commented:
а по остальному пока не смотрел, через xcode проверю вечером, но если честно все равно нифига не понял =)

s1dney commented:
что-то я уже начал сомневаться в своей правоте
во всяком случае проверь как я написал с помощью NSLog, я тоже проверю вечером

s1dney commented:
извиняюсь за необоснованный наезд, я оказался неправ, алгоритм верный
но понять его было не просто

freedomride commented:
no problem)
чужие алгоритмы бывает очень не просто понять.
еще раз спасибо за проверку)
